cfn-resources/private-endpoint/cmd/resource/steps/privateendpoint/private_endpoint.go

```diff
@@ -17,57 +18,132 @@ const (
+				peCallback.SubnetID = valStr
+			case "InterfaceEndpointID":
+				peCallback.InterfaceEndpointID = valStr
+			case "Status":
+				peCallback.Status = valStr
+			}
+		}
+		ret[i] = peCallback
+	}
+
+	s.PrivateEndpoints = ret
+
 	return nil
 }
 
-func Create(mongodbClient *mongodbatlas.Client, groupID string, interfaceEndpointID string, endpointServiceID string) handler.ProgressEvent {
-	interfaceEndpointRequest := &mongodbatlas.InterfaceEndpointConnection{
-		ID: interfaceEndpointID,
-	}
+func GetCallback(privateEndpointInput []AtlasPrivateEndpointInput, endpointServiceID string, state constants.EventStatus) (map[string]interface{}, error) {
+	endpointCallBacks := make([]AtlasPrivateEndpointCallBack, len(privateEndpointInput))
 
-	_, response, err := mongodbClient.PrivateEndpoints.AddOnePrivateEndpoint(context.Background(),
-		groupID,
-		ProviderName,
-		endpointServiceID,
-		interfaceEndpointRequest)
-	if err != nil {
-		return progress_events.GetFailedEventByResponse(fmt.Sprintf("Error creating resource : %s", err.Error()),
-			response.Response)
+	for i, pe := range privateEndpointInput {
+		callBack := AtlasPrivateEndpointCallBack{
+			VpcID:               pe.VpcID,
+			SubnetID:            pe.SubnetID,
+			InterfaceEndpointID: pe.InterfaceEndpointID,
+		}
+
+		if pe.Status != nil {
+			callBack.Status = *pe.Status
+		}
+
+		endpointCallBacks[i] = callBack
 	}
 
 	callBackContext := privateEndpointCreationCallBackContext{
-		StateName:   constants.CreatingPrivateEndpoint,
-		ID:          endpointServiceID,
-		InterfaceID: interfaceEndpointID,
+		StateName:        state,
+		ID:               endpointServiceID,
+		PrivateEndpoints: endpointCallBacks,
 	}
 
+	return getMapFromCallBackContext(callBackContext)
+}
+
+func getMapFromCallBackContext(callBackContext privateEndpointCreationCallBackContext) (map[string]interface{}, error) {
 	var callBackMap map[string]interface{}
 	data, _ := json.Marshal(callBackContext)
-	err = json.Unmarshal(data, &callBackMap)
+	err := json.Unmarshal(data, &callBackMap)
+
+	return callBackMap, err
+}
+
+func Create(mongodbClient *mongodbatlas.Client, groupID string, privateEndpointInput []AtlasPrivateEndpointInput, endpointServiceID string) handler.ProgressEvent {
+	for _, endpoint := range privateEndpointInput {
+		interfaceEndpointRequest := &mongodbatlas.InterfaceEndpointConnection{
+			ID: endpoint.InterfaceEndpointID,
+		}
+
+		_, response, err := mongodbClient.PrivateEndpoints.AddOnePrivateEndpoint(context.Background(),
+			groupID,
+			ProviderName,
+			endpointServiceID,
+			interfaceEndpointRequest)
+		if err != nil {
+			return progress_events.GetFailedEventByResponse(fmt.Sprintf("Error creating resource : %s", err.Error()),
+				response.Response)
+		}
+	}
+
+	for i := range privateEndpointInput {
+		status := StatusInitiating
+		privateEndpointInput[i].Status = &status
+	}
+
+	callBackMap, err := GetCallback(privateEndpointInput, endpointServiceID, constants.CreatingPrivateEndpoint)
```

Comment:
**Create** should only create the endpoints. Checking if the endpoints are up and running should be done in another function.

Reply:
create is only creating endpoints, the GetCallback function is creating a result structure, the create function is returning that struct in a progress event, so it can be used as an input in the asycrounous process to validate if the endpoints are running. but that validation is not done here, is done in the next steps

Reply:
Thanks for the clarification!